File: 215-kth-largest-element-in-an-array/kth-largest-element-in-an-array.java
class Solution {
    public int findKthLargest(int[] nums, int k) {
        // PriorityQueue<Integer> pq = new PriorityQueue<>();
        // for (int i = 0; i < nums.length; i++) {
        //     pq.add(nums[i]);
        //     if (pq.size() > k)
        //         pq.poll();
        // }
        // return pq.peek();
        int[] count = new int[20001];
        for(int num : nums){
            count[num + 10000]++;
        }

        for(int i = count.length - 1; i >= 0; i--){
            if(count[i] > 0){
                k -= count[i];
                if(k <= 0) return i - 10000;
            }
        }
        return -1;
    }
}

File: 215-kth-largest-element-in-an-array/kth-largest-element-in-an-array.py
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        def quick(nums, k):
            if not nums:
                return
            pivot = random.choice(nums)
            left = [x for x in nums if (x > pivot)]
            mid = [x for x in nums if (x == pivot)]
            right = [x for x in nums if (x < pivot)]

            l, m = len(left), len(mid)
            if k <= l:
                return self.findKthLargest(left, k)
            elif k > l + m:
                return self.findKthLargest(right, k - l - m)
            else:
                return mid[0]

        return quick(nums, k)
